### PR TITLE
Public API to retrieve Var bound expressions

### LINKED ISSUE #2160: Add API/Getter for Variable Bounds Expressions
## Summary

API/Getter for variable bounds expression to circumvent use of the protected `._lb` and `._ub` attributes of variable objects.

### Rationale

The module `pyomo.contrib.pyros` may require access to variable bounds expressions to determine whether these bounds are contingent on uncertain parameters, and consequently, to construct separation problem objective expressions which contain variable objects meant to simulate these uncertain parameters (as opposed to just the numerical values of the parameters). Currently, the only known way to access these variable bound expressions is through the protected `._lb` and `._ub` attributes.

### Description

See summary.

### PR DESCRIPTION
**Pull request by jsiirola**
## Summary/Motivation:
#2159 needs a  public API for retrieving the expressions for variable lower and upper bounds.  This PR adds `lower` and `upper` properties to `_VarData` to retrieve expressions for the current variable bounds.  The API is meant to mirror the `lb` / `lower` API in `Constraint`, where `lb` returns the current value and `lower` returns the expression.

To accomplish this, this PR also
- add `MaxExpression` and `MinExpression` to the expression system (so the expression can capture the tighter of the domain and the bound expression)
- makes `lb`, `ub`, `lower`, `upper`, and `bounds` all settable (to mirror behavior in kernel)
- restricts bounds expressions to being `not is_potentially_variable` (previously, Var only required bounds to be fixed)
- adds warning when declaring `RangeSet` objects with non-constant bounds (e.g., using a mutable Param now yields a warning)
- fix implementation of `_VarData.free()` so that free works correctly for derived classes that override `unfix()` 

## Changes proposed in this PR:
- (see above)

### Legal Acknowledgement

By contributing to this software project, I have read the [contribution guide](https://pyomo.readthedocs.io/en/stable/contribution_guide.html) and agree to the following terms and conditions for my contribution:

1. I agree my contributions are submitted under the BSD license.
2. I represent I am authorized to make the contributions and grant the license. If my employer has rights to intellectual property that includes these contributions, I represent that I have received permission to make contributions and grant the required license on behalf of that employer.

### CODE REVIEW

File: pyomo/core/base/var.py [hunk 1]
```diff
@@ -244,13 +280,13 @@ def domain(self):
         raise NotImplementedError
 
     @property
-    def lb(self):
-        """Return the lower bound for this variable."""
+    def lower(self):
+        """Return an expression for the vaiable lower bound."""
```
Comment (mrmundt) [inline pyomo/core/base/var.py:284]: Typo: `variable`

File: pyomo/core/base/var.py [hunk 2]
```diff
@@ -82,24 +84,58 @@ def __init__(self, component=None):
...
     def bounds(self, val):
-        raise AttributeError("Assignment not allowed. Use the setub and setlb methods")
+        self.lower, self.upper = val
+
+    @property
+    def lb(self):
+        """Return the numeric value of the variable lower bound."""
+        return value(self.lower)
+    @lb.setter
+    def lb(self, val):
```
Comment (mrmundt) [inline pyomo/core/base/var.py:129]: If tests for `lb.setter` and `ub.setter` are relatively simple to write, we may want to add them.

File: pyomo/core/base/var.py [hunk 3]
```diff
@@ -411,30 +434,52 @@ def domain(self, domain):
...
             return dlb
         elif dlb is None:
-            return value(self._lb)
-        return max(value(self._lb), dlb)
-    @lb.setter
-    def lb(self, val):
-        raise AttributeError("Assignment not allowed. Use the setlb method")
+            return self._lb
+        # This is guaranteed by _process_bound():
+        # assert not is_potentially_variable(self._lb)
```
Comment (mrmundt) [inline pyomo/core/base/var.py:454]: You made a comment to me earlier about "commented out code causing mayhem", so I'll regurgitate that back to you here.

File: pyomo/core/base/var.py [hunk 4]
```diff
@@ -411,30 +434,52 @@ def domain(self, domain):
...
             return dub
         elif dub is None:
-            return value(self._ub)
-        return min(value(self._ub), dub)
-    @ub.setter
-    def ub(self, val):
-        raise AttributeError("Assignment not allowed. Use the setub method")
+            return self._ub
+        # This is guaranteed by _process_bound():
+        # assert not is_potentially_variable(self._lb)
```
Comment (mrmundt) [inline pyomo/core/base/var.py:478]: You made a comment to me earlier about "commented out code causing mayhem", so I'll regurgitate that back to you here.

**Review by mrmundt — CHANGES_REQUESTED** [review]

**Review by mrmundt — APPROVED** [review]

File: pyomo/core/base/var.py [hunk 5]
```diff
@@ -82,24 +84,58 @@ def __init__(self, component=None):
...
 
     def has_ub(self):
         """Returns :const:`False` when the upper bound is
         :const:`None` or positive infinity"""
-        ub = self.ub
-        return (ub is not None) and \
-            (value(ub) != float('inf'))
+        return self.ub not in _no_upper_bound
+
+    # TODO: deprecate this?  Properties are generally preferred over "set*()"
```
Comment (emma58) [inline pyomo/core/base/var.py:94]: I would be in favor of this idea...

File: pyomo/core/base/var.py (hunk 2 above)
Comment (emma58) [reply pyomo/core/base/var.py:129]: Also, I think that the docstrings for `_VarData` and `_GeneralVarData` needs to be updated now that these *do* have setters.

**Review by emma58 — APPROVED** [review]
I think a couple of the docstrings now lie, but otherwise this looks good to me!

**Pull request merged**